sklearn/cluster/_spectral.py

```diff
@@ -744,7 +744,7 @@ def fit(self, X, y=None):
                 params["coef0"] = self.coef0
             self.affinity_matrix_ = pairwise_kernels(
                 X, metric=self.affinity, filter_params=True, **params
-            )
+            ).astype(X.dtype, copy=False)
```

Comment:
I think we should work on making `pairwise_kernels` work efficiently with `float32` data first.

Since `affinity="rbf"` is the default, flagging `SpectralClustering` as `dtype` preserving without that prerequisite would be misleading to our users: there would be very few performance or peak memory usage gains in passing `float32` data to this estimator when using the default hyper-params.

Reply:
BTW we also need a test that checks the dtype of `affinity_matrix_` when the `affinity` param is the str name of a kernel function and another such assertion in  a test that covers the case when affinity is `nearest_neighbors`.

Reply:
We also need to have `ArgKmin` and `RadiusNeighbors` preserve dtypes because it is also used when `affinity="precomputed_nearest_neighbors"`.

Let's turn this PR as a draft for now.